src/Workspaces/CSharp/Portable/VirtualChars/CSharpVirtualCharService.cs

```diff
@@ -0,0 +1,255 @@
+
+                // Surrogate characters aren't supported here.
+                if (uintChar >= 0x00010000)
+                {
+                    // This is possible.  It's a legal C# escape, but we don't support it here because it
+                    // would need two chars to encode.
+                    return false;
+                }
+
+                result.Add(new VirtualChar((char)uintChar, new TextSpan(startIndex + offset, 2 + 8)));
+                return true;
+            }
+            else if (character == 'u')
+            {
+                // 4 character escape representing one char.
+
+                var intChar = 0;
+                if (!IsHexDigit(tokenText[index]))
+                {
+                    Debug.Assert(false, "This should not be reachable as long as the compiler added no diagnostics.");
+                    return false;
+                }
+
+                var endIndex = index + 1;
+                for (var i = 0; i < 4; i++)
+                {
+                    var ch2 = tokenText[index + i];
+                    if (!IsHexDigit(ch2))
+                    {
+                        Debug.Assert(false, "This should not be reachable as long as the compiler added no diagnostics.");
+                        return false;
+                    }
+
+                    intChar = (intChar << 4) + HexValue(ch2);
+                    endIndex++;
+                }
+
+                character = (char)intChar;
+                result.Add(new VirtualChar(character, new TextSpan(startIndex + offset, 2 + 4)));
+                return true;
+            }
+            else
+            {
+                Debug.Assert(character == 'x');
+                // Variable length (up to 4 chars) hexidecimal escape.
+
+                var intChar = 0;
+                if (!IsHexDigit(tokenText[index]))
+                {
+                    Debug.Assert(false, "This should not be reachable as long as the compiler added no diagnostics.");
+                    return false;
+                }
+
+                var endIndex = index;
+                for (var i = 0; i < 4; i++)
+                {
+                    var ch2 = tokenText[index + i];
+                    if (!IsHexDigit(ch2))
+                    {
+                        // This is possible.  These escape sequences are variable length.
+                        break;
+                    }
+
+                    intChar = (intChar << 4) + HexValue(ch2);
+                    endIndex++;
+                }
+
+                character = (char)intChar;
+                result.Add(new VirtualChar(character, TextSpan.FromBounds(startIndex + offset, endIndex + offset)));
+                return true;
+            }
+        }
+
+        private static int HexValue(char c)
+        {
+            Debug.Assert(IsHexDigit(c));
+            return (c >= '0' && c <= '9') ? c - '0' : (c & 0xdf) - 'A' + 10;
+        }
+
+        private static bool IsHexDigit(char c)
```

Comment:
Doesn't cover fullwidth characters that are valid hex character in VB.net

Reply:
This is the C# impl.

Reply:
My apologise, thought it was in the base class used for both.

Reply:
#wontfix